src/app/clusters/time-synchronization-server/time-synchronization-cluster.cpp

```diff
@@ -238,49 +204,265 @@ static bool emitMissingTrustedTimeSourceEvent(EndpointId ep)
+{
+    auto timeSynchronization = FindClusterOnEndpoint(endpoint);
+    VerifyOrReturnValue(timeSynchronization != nullptr, CHIP_ERROR_INVALID_ARGUMENT);
+
+    CHIP_ERROR err = encoder.EncodeList([&timeSynchronization](const auto & encod) -> CHIP_ERROR {
+        const auto & dstList = timeSynchronization->GetDSTOffset();
+        for (const auto & dstOffset : dstList)
+        {
+            ReturnErrorOnFailure(encod.Encode(dstOffset));
+        }
+
+        return CHIP_NO_ERROR;
+    });
+
+    return err;
+}
+
+CHIP_ERROR TimeSynchronizationCluster::ReadLocalTime(EndpointId endpoint, AttributeValueEncoder & encoder)
+{
+    DataModel::Nullable<uint64_t> localTime;
+
+    auto timeSynchronization = FindClusterOnEndpoint(endpoint);
+    VerifyOrReturnValue(timeSynchronization != nullptr, CHIP_ERROR_INVALID_ARGUMENT);
+
+    CHIP_ERROR err = timeSynchronization->GetLocalTime(endpoint, localTime);
+    if (err == CHIP_NO_ERROR)
+    {
+        err = encoder.Encode(localTime);
+    }
+    return err;
+}
+
+CHIP_ERROR TimeSynchronizationCluster::Attributes(const ConcreteClusterPath & path,
+                                                  ReadOnlyBufferBuilder<DataModel::AttributeEntry> & builder)
+{
+    AttributeListBuilder listBuilder(builder);
+    return listBuilder.Append(Span(TimeSynchronization::Attributes::kMandatoryMetadata), {});
+}
+
+DataModel::ActionReturnStatus TimeSynchronizationCluster::ReadAttribute(const DataModel::ReadAttributeRequest & request,
+                                                                        AttributeValueEncoder & encoder)
+{
+    switch (request.path.mAttributeId)
+    {
+    case UTCTime::Id: {
+        System::Clock::Microseconds64 utcTimeUnix;
+        uint64_t chipEpochTime;
+        auto timeSynchronization = FindClusterOnEndpoint(request.path.mEndpointId);
+        VerifyOrReturnValue(timeSynchronization != nullptr, CHIP_ERROR_INVALID_ARGUMENT);
+        // This return allows us to simulate no internal time for testing purposes
+        // This will be set once we receive a good time either from the delegate or via a command
+        if (timeSynchronization->GetGranularity() == GranularityEnum::kNoTimeGranularity)
+        {
+            return encoder.EncodeNull();
+        }
+        VerifyOrReturnError(System::SystemClock().GetClock_RealTime(utcTimeUnix) == CHIP_NO_ERROR, encoder.EncodeNull());
+        VerifyOrReturnError(UnixEpochToChipEpochMicros(utcTimeUnix.count(), chipEpochTime), encoder.EncodeNull());
+        return encoder.Encode(chipEpochTime);
+    }
+    case Granularity::Id: {
+        auto timeSynchronization = FindClusterOnEndpoint(request.path.mEndpointId);
+        VerifyOrReturnValue(timeSynchronization != nullptr, CHIP_ERROR_INVALID_ARGUMENT);
+        return encoder.Encode(timeSynchronization->GetGranularity());
+    }
+    case TrustedTimeSource::Id:
+        return ReadTrustedTimeSource(request.path.mEndpointId, encoder);
```

Comment:
passing endpointID on each of these feels off ... this is guaranteed to be for `this` cluster as a call, so switching it to a different one feels like some extra indirection.